src/main/drivers/accgyro/accgyro_spi_icm456xx.c

```diff
@@ -490,16 +521,78 @@ bool icm456xxSpiAccDetect(accDev_t *acc)
 
 bool icm456xxGyroReadSPI(gyroDev_t *gyro)
 {
-    uint8_t raw[6] = {0};
+    int16_t *gyroData = (int16_t *)gyro->dev.rxBuf;
+    switch (gyro->gyroModeSPI) {
+    case GYRO_EXTI_INIT:
+    {
+        // Initialise the tx buffer to all 0xff
+        memset(gyro->dev.txBuf, 0xff, 12);
+
+        // Check that minimum number of interrupts have been detected
+
+        // We need some offset from the gyro interrupts to ensure sampling after the interrupt
+        gyro->gyroDmaMaxDuration = 5;
+        if (gyro->detectedEXTI > GYRO_EXTI_DETECT_THRESHOLD) {
+#ifdef USE_DMA
+            if (spiUseDMA(&gyro->dev)) {
+                gyro->dev.callbackArg = (uint32_t)gyro;
+                gyro->dev.txBuf[0] = ICM456XX_GYRO_DATA_X1_UI | 0x80;
+                gyro->segments[0].len = 12;
+                gyro->segments[0].callback = mpuIntCallback;
+                gyro->segments[0].u.buffers.txData = gyro->dev.txBuf;
+                gyro->segments[0].u.buffers.rxData = &gyro->dev.rxBuf[1];
+                gyro->segments[0].negateCS = true;
+                gyro->gyroModeSPI = GYRO_EXTI_INT_DMA;
+            } else
+#endif
+            {
+                // Interrupts are present, but no DMA
+                gyro->gyroModeSPI = GYRO_EXTI_INT;
+            }
+        } else {
+            gyro->gyroModeSPI = GYRO_EXTI_NO_INT;
+        }
+        break;
+    }
 
-    const bool ack = spiReadRegMskBufRB(&gyro->dev, ICM456XX_GYRO_DATA_X1_UI, raw, 6);
-    if (!ack) {
-        return false;
+    case GYRO_EXTI_INT:
+    case GYRO_EXTI_NO_INT:
+    {
+        gyro->dev.txBuf[0] = ICM456XX_GYRO_DATA_X1_UI | 0x80;
+
+        busSegment_t segments[] = {
+                {.u.buffers = {NULL, NULL}, 6, true, NULL},
+                {.u.link = {NULL, NULL}, 0, true, NULL},
+        };
+        segments[0].u.buffers.txData = gyro->dev.txBuf;
+        segments[0].u.buffers.rxData = &gyro->dev.rxBuf[1];
+
+        spiSequence(&gyro->dev, &segments[0]);
+
+        // Wait for completion
+        spiWait(&gyro->dev);
+
+        gyro->gyroADCRaw[X] = gyroData[1];
```

Comment:
This is gyro->dev.rxBuf + 2

Reply:
Why + 2 ?
Guro and Accel data are read correctly
<img width="1152" alt="Знімок екрана 2025-05-02 о 20 08 13" src="https://github.com/user-attachments/assets/fa08e472-f904-42b1-a0cf-0d5ce370284a" />

Reply:
gyroData is `(uint16_t*)` so you get 2byte offset from `[1]`. 
Even with correct offset, it can result in unaligned read (rxBuf is aligned). so it is undefined behavior in C. It is very unlikely to cause problems on cortex M3 and higher, but better avoid such UB

Reply:
Agree. Сan fill in X Y Z explicitly. 
Like that
```C
gyro->gyroADCRaw[X] = (int16_t)((gyro->dev.rxBuf[1] << 8) | gyro->dev.rxBuf[0]);
gyro->gyroADCRaw[Y] = (int16_t)((gyro->dev.rxBuf[3] << 8) | gyro->dev.rxBuf[2]);
gyro->gyroADCRaw[Z] = (int16_t)((gyro->dev.rxBuf[5] << 8) | gyro->dev.rxBuf[4]);
```